### PR TITLE
Add more infrastructure for source/boundary condition handling.

### PR DESCRIPTION
**Pull request by simonbyrne**
This lets you feed in tuples to source or boundary conditions:
- for source terms they will apply additively
- for boundary conditions, they will be indexed by `bctype`

### CODE REVIEW

**Comment by simonbyrne** [comment]
bors try

File: test/DGmethods/compressible_Navier_Stokes/rising_bubble-model.jl [hunk 1]
```diff
@@ -110,7 +103,9 @@ function run(mpicomm, ArrayType,
                      SmagorinskyLilly{DF}(C_smag),
                      EquilMoist(), 
                      NoRadiation(),
-                     source_geopot!, NoFluxBC(), initialise_rising_bubble!)
+                     Gravity(),
```
Comment (charleskawczynski) [inline test/DGmethods/compressible_Navier_Stokes/rising_bubble-model.jl]: Shouldn't this by a tuple? (e.g., `(Gravity(), )`)

Comment (simonbyrne) [reply test/DGmethods/compressible_Navier_Stokes/rising_bubble-model.jl]: The idea is that you can specify either a single source, or a tuple of sources.
